File: src/C#/NStreamCom/Packet.cs
using System;
using System.IO;

namespace NStreamCom
{
    public class Packet
    {
        public ushort ID;

        public ushort MessageSize;

        public byte[] Data;

        public int StreamBytesSize
        {
            get => Data.Length + Protocol.MetadataSize;
        }

        public Packet(ushort id, ushort messageSize, byte[] data)
        {
            ID = id;
            MessageSize = messageSize;
            Data = data;
        }

        public Packet(ushort id, ushort messageSize, Stream data)
        {
            ID = id;
            MessageSize = messageSize;
            Data = data.ReadAllBytes();
        }

        public Packet(byte[] data)
        {
            Protocol.VerifyBytes(data);
            ID = BitConverter.ToUInt16(data, 0);
            MessageSize = BitConverter.ToUInt16(data, 2);
            Data = new byte[BitConverter.ToUInt16(data, 4)];
            Array.Copy(data, 6, Data, 0, Data.Length);
        }

        public Packet(Stream stream)
            : this(stream.ReadAllBytes())
        {
        }

        public byte[] GetStreamBytes()
        {
            byte[] streamBytes = new byte[StreamBytesSize];
            Array.Copy(BitConverter.GetBytes(ID), 0, streamBytes, 0, 2);
            Array.Copy(BitConverter.GetBytes(MessageSize), 0, streamBytes, 2, 2);
            Array.Copy(BitConverter.GetBytes((ushort)Data.Length), 0, streamBytes, 4, 2);
            Array.Copy(Data, 0, streamBytes, 6, Data.Length);
            return streamBytes;
        }
    }
}

File: src/Python/NStreamCom/Packet.py
STREAM_PROTOCOL_SIZE = 6

class PacketSizeDataMismatch(Exception):
    def __init__(self, message : str):
        super.__init__(PacketSizeDataMismatch, self, message)

def verify_bytes(data : list):
    if len(data) < STREAM_PROTOCOL_SIZE:
        raise PacketSizeDataMismatch('Length of list was less than minimum')
    
    message_size = int.from_bytes(data[4:7], 'little')

    if len(data) - STREAM_PROTOCOL_SIZE != message_size:
        raise PacketSizeDataMismatch("Length of list did not match its size with protocol")
    
    return message_size

class Packet:
    def __init__(self, id : int = 0, message_size : int = 0, data = None):
        self.id = id
        self.message_size = message_size
        self.data = data

    def __init__(self, data):
        self.message_size = verify_bytes(data)
        self.id = int.from_bytes(data[0:2], 'little')
        self.data = data[STREAM_PROTOCOL_SIZE:]

    def get_stream_bytes(self) -> list:
        stream_bytes = []
        stream_bytes +=  int.to_bytes(self.id, 2, 'little')
        stream_bytes += int.to_bytes(self.message_size, 2, 'little')
        stream_bytes += int.to_bytes(len(self.data), 'little')
        if self.data is not None:
            stream_bytes += self.data
        return stream_bytes